include/aos/iam/certmodule.hpp

```diff
@@ -0,0 +1,386 @@
+         * @param certInfo certificate info.
+         * @return void.
+         */
+        virtual void OnCertAdded(const String& certType, const CertInfo& certInfo) = 0;
+
+        /**
+         * Called whenever certificate is removed from the module.
+         *
+         * @param certType certificate type.
+         * @param certURL certificate URL.
+         * @return void.
+         */
+        virtual void OnCertRemoved(const String& certType, const String& certURL) = 0;
+
+        /**
+         * Destroys object instance.
+         */
+        virtual ~CertListener() = default;
+    };
+
+    /**
+     * Creates a new object instance.
+     *
+     * @param certType certificate type.
+     * @param config module config.
+     */
+    CertModule(const String& certType, const ModuleConfig& config);
+
+    /**
+     * Initializes certificate module.
+     *
+     * @param x509Provider provider of x509 certificates, certificate requests, keys.
+     * @param secureStorage a reference to secure storage.
+     * @return Error.
+     */
+    Error Init(crypto::x509::ProviderItf& x509Provider, HSMItf& hsm);
+
+    /**
+     * Sets certificate listener.
+     *
+     * @param listener .
+     * @return void.
+     */
+    void SetListener(CertListener& listener);
+
+    /**
+     * Returns IAM module certificate type.
+     *
+     * @return const String&.
+     */
+    const String& GetCertType() const;
+
+    /**
+     * Returns valid certificates.
+     *
+     * @return const Array<CertInfo>&.
+     */
+    const Array<CertInfo>& GetValidCerts() const;
+
+    /**
+     * Owns the module.
+     *
+     * @param password certficate password.
+     * @return Error.
+     */
+    Error SetOwner(const String& password);
+
+    /**
+     * Removes all module certificates.
+     *
+     * @return Error.
+     */
+    Error Clear();
+
+    /**
+     * Generates private key.
+     *
+     * @param password owner password.
+     * @return RetWithError<const crypto::PrivateKey*> .
+     */
+    RetWithError<const crypto::PrivateKey*> CreateKey(const String& password);
+
+    /**
+     * Creates certificate request.
+     *
+     * @param subject certificate subject.
+     * @param privKey private key.
+     * @param[out] csr result csr.
+     * @return Error.
+     */
+    Error CreateCSR(const Array<uint8_t>& subject, const crypto::PrivateKey& privKey, crypto::x509::CSR& csr);
+
+    /**
+     * Applies certificate to a module.
+     *
+     * @param pemCert certificate chain in PEM format.
+     * @param[out] info result certificate information.
+     * @returns Error.
+     */
+    Error ApplyCert(const Array<uint8_t>& pemCert, CertInfo& info);
+
+    /**
+     * Creates a self signed certificate.
+     *
+     * @param password owner password.
+     * @returns Error.
+     */
+    Error CreateSelfSignedCert(const String& password);
+
+private:
+    static constexpr auto cCertChainSize = AOS_CONFIG_CERTHANDLER_CERTS_CHAIN_SIZE;
```

Comment:
Could you move all private members to implementation commit? Actually, all private members should be defined in the commit they are used. Otherwise, it is hard to review.

Reply:
Basically that's difficult because of two reasons:  git gui cant unstage lines for a new file, an alternative command line approach is a way more difficult to use, also it would involve some modification to dependent config values.

Reply:
Actually, it is quite easy to achieve using interactive rebase: cut it in this commit and insert in another one. Do not understand about config values. You can keep config values in this commit.